modules/features/player/src/main/java/androidx/media3/extractor/text/webvtt/WebvttCueParser.java

```diff
@@ -0,0 +1,998 @@
+            startTagStack.push(StartTag.buildStartTag(fullTagExpression, spannedText.length()));
+          }
+          break;
+        case CHAR_AMPERSAND:
+          int semiColonEndIndex = markup.indexOf(CHAR_SEMI_COLON, pos + 1);
+          int spaceEndIndex = markup.indexOf(CHAR_SPACE, pos + 1);
+          int entityEndIndex =
+                  semiColonEndIndex == -1
+                          ? spaceEndIndex
+                          : (spaceEndIndex == -1
+                          ? semiColonEndIndex
+                          : min(semiColonEndIndex, spaceEndIndex));
+          if (entityEndIndex != -1) {
+            applyEntity(markup.substring(pos + 1, entityEndIndex), spannedText);
+            if (entityEndIndex == spaceEndIndex) {
+              spannedText.append(" ");
+            }
+            pos = entityEndIndex + 1;
+          } else {
+            spannedText.append(curr);
+            pos++;
+          }
+          break;
+        default:
+          spannedText.append(curr);
+          pos++;
+          break;
+      }
+    }
+    // apply unclosed tags
+    while (!startTagStack.isEmpty()) {
+      applySpansForTag(id, startTagStack.pop(), nestedElements, spannedText, styles);
+    }
+    applySpansForTag(
+            id,
+            StartTag.buildWholeCueVirtualTag(),
+            /* nestedElements= */ Collections.emptyList(),
+            spannedText,
+            styles);
+    return SpannedString.valueOf(spannedText);
+  }
+
+  // Internal methods
+
+  @Nullable
+  private static WebvttCueInfo parseCue(
+          @Nullable String id,
+          Matcher cueHeaderMatcher,
+          ParsableByteArray webvttData,
+          List<WebvttCssStyle> styles) {
+    WebvttCueInfoBuilder builder = new WebvttCueInfoBuilder();
+    try {
+      // Parse the cue start and end times.
+      builder.startTimeUs =
+              WebvttParserUtil.parseTimestampUs(Assertions.checkNotNull(cueHeaderMatcher.group(1)));
+      builder.endTimeUs =
+              WebvttParserUtil.parseTimestampUs(Assertions.checkNotNull(cueHeaderMatcher.group(2)));
+    } catch (NumberFormatException e) {
+      Log.w(TAG, "Skipping cue with bad header: " + cueHeaderMatcher.group());
+      return null;
+    }
+
+    parseCueSettingsList(Assertions.checkNotNull(cueHeaderMatcher.group(3)), builder);
+
+    // Parse the cue text.
+    StringBuilder textBuilder = new StringBuilder();
+    for (String line = webvttData.readLine();
+         !TextUtils.isEmpty(line);
+         line = webvttData.readLine()) {
+      if (textBuilder.length() > 0) {
+        textBuilder.append("\n");
+      }
+      textBuilder.append(line.trim());
+    }
+    String markup = textBuilder.toString();
+    builder.text = parseCueText(id, markup, styles);
+    // parseCueText strips tags like TAG_VOICE "v" and set it on the cue text. This sets original markup wih tags in cue to allow extracting it in the application.
+    builder.markup = markup;
```

Comment:
`Cue.text` for vtt format parser strips out tags like TAG_VOICE "v" without setting it on the cue. This sets original markup with tags on the cue to allow extracting any tag in the application in 7548a3f.

I copied original source code for `WebvttCueParser` and `Cue` to make the change as the classes were final and could not be subclassed. Curious to know if there's a better way to handle this.

Reply:
Could you report the issue in the `media` repository? Hopefully they can add needed APIs and we can remove copied classes in the future. If they agree that this is something they could add we could make a PR to their repo.

Reply:
Asked here: https://github.com/androidx/media/issues/1632